src/AESmodes.cpp

```diff
@@ -5,268 +5,811 @@
-  }
-
-  std::array<unsigned char, 16> outputPadded;
-  decrypt(block, outputPadded, key);
-
-  for (std::size_t i = 0; i < 16; i++) {
-    outputPadded[i] ^= IV[i];
-    output.push_back(outputPadded[i]); 
-  }
-
-  // Loop over the number of subsequent blocks
-  for (std::size_t i = 1; i < inputSize / 16; i++) {
-
-    // Loop over block size and fill each block
-    std::array<unsigned char, 16> block;
-    for (std::size_t j = 0; j < 16; j++) {
-      block[j] = input[j+(i*16)];
+void incrementCounter(std::array<unsigned char, NUM_BYTES> &counter, int numCounterBytes) noexcept(false) {
+    for (int i = NUM_BYTES - 1; i >= numCounterBytes; i--) {
+        //Increment the current byte
+        // Secure coding: CTR50-CPP. Guarantee that container indices and iterators are within the valid range
+        counter.at(i) = counter.at(i) + 1;
+        //If the byte did not overflow to zero, then stop
+        //Otherwise continue until an overflow does not happen
+        if (counter.at(i) != 0)
+            break;
     }
+}
 
-    // Decrypt each block
-    std::array<unsigned char, 16> outputPadded;
-    decrypt(block, outputPadded, key);
+/**
+  cipher with CTR mode
+    Guaranteed no exceptions by:
+    handling all exceptions per ERR51-CPP
+        Related: Honoring exception specifications, all exceptions will be caught per ERR55-CPP
+    not throwing exceptions across execution boundaries (library to application) per ERR59-CPP
+    Guaranteeing Strong exception safety per ERR56-CPP
+        Program state will not be modified
+            Input, key, and nonce are constant and output vector is cleared when catching an exception
+  @param input: vector of hex values representing the plaintext
+  @param output: vector of hex values representing ciphertext (with padding)
+  @param key: vector of hex values representing key to use
+  @param nonce: a NUM_BYTES/2 (8) byte random block for the counter
+  @return True on success
+*/
+bool encrypt_ctr(const std::vector<unsigned char> &input, std::vector<unsigned char> &output,
+                 const std::vector<unsigned char> &key,
+                 const std::array<unsigned char, NUM_BYTES / 2> &nonce) noexcept(true) {
+    try {
+        // Calculate padding length, then copy input array and padding into plaintext
+        const std::size_t inputSize = input.size();
+        const std::size_t padLength = NUM_BYTES - (inputSize % NUM_BYTES);
+        const std::size_t plaintextLength = inputSize + padLength;
+
+        // Plaintext accommodates both the input and the necessary padding
+        std::vector<unsigned char> plaintext;
+        plaintext.reserve(plaintextLength);
+        plaintext = input;
+
+        // TODO: investigate whether PKCS#7 is the best choice for padding (padding oracle attack)
+        for (std::size_t i = 0; i < padLength; i++) {
+            // PKCS#7 padding (source: https://www.ibm.com/docs/en/zos/2.1.0?topic=rules-pkcs-padding-method)
```

Comment:
Should ctr mode be using padding? I thought that's just for ecb cbc and cfb

Reply:
ECB, CBC and CFB require padding and the others don’t, but we can still use padding to figure out the size of the decrypted plaintext.